### PR TITLE
support reading/writing fds

### PR DESCRIPTION
**Pull request by cjihrig**
This PR makes `fd_read()` and `fd_write()` work (with an included test). It also makes the existing `test/wasi/test-wasi-binding.js` functional (cc: @danbev). Finally, this PR makes the preopen directories configurable.

### CODE REVIEW

File: test/wasi/c/read_file.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include <assert.h>
+#include <stdio.h>
+
+int main() {
+  FILE* file = fopen("/sandbox/input.txt", "r");
+  assert(file != NULL);
+
+  char c = fgetc(file);
+  while (c != EOF) {
+    assert(fputc(c, stdout) != EOF);
```
Comment (legendecas) [inline test/wasi/c/read_file.c]: As assert() macro may being removed at compile time with -DNDEBUG, expressions that have side effects should not be put in `assert()`?

**Review by danbev — APPROVED** [review]
👍 With the comment from @legendecas. 
Was very nice to see `test-wasi-binding.js` working 😄, thanks!

File: test/wasi/c/read_file.c (hunk 1 above)
Comment (cjihrig) [reply test/wasi/c/read_file.c]: Updated. Thanks.

**Comment by Fishrock123** [comment]
It'd be nice if we could add a WAST version of the test file(s), so they could be read by humans and changed/recompiled?

**Comment by cjihrig** [comment]
@Fishrock123 the C source files are in `test/wasi/c/*.c`. There is a `Makefile` in `test/wasi` (thanks to Dan) that recompiles them. I think it's important to make sure that the WASI implementation actually works with the WASI libc.

However, there is one test currently (`test/wasi/test-wasi-binding.js`, also added by Dan) that is based on a WAT file (`test/fixtures/wasi/simple-wasi.wat`).

More tests are certainly welcome!

**Pull request merged**